Add CellAssign predictions to cell type annotation workflow

allyhawkins opened the pull request:
Closes #303 

This PR adds a process for obtaining `CellAssign` predictions to the main cell type annotation workflow. 

- I added this process into the current `classify_celltypes.nf` process where we now get cell type annotations using both `SingleR` and `CellAssign`. Right now for `CellAssign` I only output the predictions and plan to add one more process to merge in the predictions from `CellAssign` to the annotated rds file with `SingleR` annotations. 
- This does rely on the assumption that every project will have both a `SingleR` reference and a `CellAssign` reference. I think that's okay, because I don't see any scenarios as of yet that we wouldn't have them. 
- I used the highest memory and cpu label that we have. During testing, this was immediately failing due to memory errors, but is running on the third retry. I'm using the max to try and eliminate the failing and re-submitting as much as possible.
- Also see [AlexsLemonade/ScPCA-admin#](https://github.com/AlexsLemonade/ScPCA-admin/pull/604) for the changes to the metadata file that I made. There I added columns for the `SingleR` ref file and `CellAssign` ref file and am directly using them in the workflow rather than creating the paths. 

~~**Note:** I am still running a test to make sure that this completes fully, but it's been running for some time so I think most of the errors have been accounted for. I'll request a review once I ensure this is able to run to completion.~~

Edit: The test completed successfully, but it did take ~16 hours as a warning for reviewers that would like to test it. 😬

sjspielman commented on add-celltypes.nf:
```diff
@@ -52,10 +52,12 @@ workflow {
-    .map{meta -> tuple(meta,
-                       file("${params.results_dir}/${meta.project_id}/${meta.sample_id}/${meta.library_id}_processed.rds")
-                       )}
+      // tuple of meta, processed RDS file to use as input to singleR
```
comment needs to be updated for 2 methods now

sjspielman commented on bin/classify_cellassign.py:
```diff
@@ -0,0 +1,92 @@
+                    dest = 'input_hdf5_file',
+                    required = True,
+                    help = 'Path to HDF5 file with processed AnnData object to annotate')
+parser.add_argument('-o', '--output_predictions',
+                    dest = 'output_predictions',
+                    required = True,
+                    help = 'Path to directory to save the predictions, should end in tsv')
```
directories dont end in tsv?

sjspielman commented on bin/classify_cellassign.py:
```diff
@@ -0,0 +1,92 @@
+parser.add_argument('-r', '--reference',
+                    dest = 'reference',
+                    required = True,
+                    help = 'Path to marker by cell type reference file')
```
can you indicate in help too if this is csv or tsv? (see comments below also)

sjspielman commented on bin/classify_cellassign.py:
```diff
@@ -0,0 +1,92 @@
+    raise FileExistsError("--reference file not found.")
+
+# make sure output file path is tsv file
+if not ".tsv" in args.output_predictions:
+    raise ValueError("--output_predictions must provide a file path ending in tsv")
```
```suggestion
if not args.output_predictions.endswith(".tsv"):
    raise ValueError("--output_predictions must provide a file path ending in tsv")
```

sjspielman commented on bin/classify_cellassign.py:
```diff
@@ -0,0 +1,92 @@
+elif not file_ext.search(args.input_hdf5_file):
+    raise ValueError("--input_hdf5_file must end in either .hdf5 or .h5 and contain a processed AnnData object.")
+
+# check that marker file exists and make sure its a csv
```
wait, csv or tsv? later you parse it as tsv?

sjspielman commented on bin/classify_cellassign.py:
```diff
@@ -0,0 +1,92 @@
+    raise ValueError("--output_predictions must provide a file path ending in tsv")
+
+# read in references as marker gene tables
+ref_matrix = pd.read_csv(args.reference, sep = "\t", index_col='ensembl_id')
```
noting the reference is parsed here as tsv

sjspielman commented on bin/classify_cellassign.py:
```diff
@@ -0,0 +1,92 @@
+import anndata as adata
+import scvi
+from scvi.external import CellAssign
+from scvi import settings
```
Is this needed given line 10? On line 45 when you set the seed, you're doing `scvi.settings.seed` anyways, so I think this can go.

sjspielman commented on modules/classify-celltypes.nf:
```diff
@@ -43,10 +69,24 @@ workflow annotate_celltypes {
+
+      predict_cellassign(cellassign_input_ch)
 
     emit: classify_singleR.out
```
Would we want to emit a tuple of both outputs here?

allyhawkins replied:
I'm going to temporarily update the output to have both, but note that this will probably change as part of #394.

sjspielman reviewed (commented):
Overall this looks good! Returning these initial comments before I get to testing it out since I understand it's gonna take a while...Has `CellAssign` always taken _this_ long for us? Curious which library you used here, so I can maybe use a smaller one 😂

allyhawkins commented:
Thanks for the review @sjspielman. I updated the comments and addressed all of your feedback. I did temporarily update the output, but again I do think that will ultimately change. 

For testing I used, SCPCL000295. I will say that I tested before I upped the cpu requirement, so it only used 4 cpus. Increasing that to 12 should hopefully decrease the amount of time we need to run.

sjspielman commented:
I'll try the same library then, let's see how the change in cpu/ram affected runtime!

sjspielman reviewed (approved):
Successfully ran through the same sample at a runtime of ~7 hours (not including submission time), which is better than 16! A couple stats from this process..
 
<img width="1366" alt="Screenshot 2023-08-02 at 1 02 34 PM" src="https://github.com/AlexsLemonade/scpca-nf/assets/4701111/ab1e0071-161a-4b04-b8aa-2fa2b0ad8359">

Given next steps coming in #394, this seems good to go for this stage!
